tests/shadow/test_transforms.py

```diff
@@ -0,0 +1,440 @@
+        """Test that the state reconstruction is correct for a subset
+        of the qubits"""
+        wires_list = [[0], [0, 1]]
+
+        circuit = max_entangled_circuit(3)
+        circuit = qml.shadows.state(wires=wires_list, diffable=diffable)(circuit)
+
+        actual = circuit()
+
+        expected = [
+            np.array([[0.5, 0], [0, 0.5]]),
+            np.array([[0.5, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0.5]]),
+        ]
+
+        assert qml.math.allclose(actual[0], expected[0], atol=1e-1)
+        assert qml.math.allclose(actual[1], expected[1], atol=1e-1)
+
+
+@pytest.mark.all_interfaces
+class TestStateForwardInterfaces:
+    """Test that state reconstruction works for all interfaces"""
+
+    @pytest.mark.parametrize("interface", ["autograd", "jax", "tf", "torch"])
+    @pytest.mark.parametrize("diffable", [True, False])
+    def test_qft_state(self, interface, diffable):
+        """Test that the state reconstruction is correct for a QFT state"""
+        circuit = qft_circuit(3, interface=interface)
+        circuit = qml.shadows.state(wires=[0, 1, 2], diffable=diffable)(circuit)
+
+        actual = circuit()
+        expected = np.exp(np.arange(8) * 2j * np.pi / 8) / np.sqrt(8)
+        expected = np.outer(expected, np.conj(expected))
+
+        assert qml.math.allclose(actual, expected, atol=1e-1)
+
+
+class TestStateBackward:
+    """Test that the gradient of the state reconstruction is correct"""
+
+    @pytest.mark.autograd
+    def test_backward_autograd(self):
+        """Test the gradient of the state for the autograd interface"""
+        shadow_circuit = basic_entangler_circuit(3, shots=20000, interface="autograd")
+        shadow_circuit = qml.shadows.state(wires=[0, 1, 2], diffable=True)(shadow_circuit)
+        exact_circuit = basic_entangler_circuit_exact_state(3, "autograd")
+
+        # make rotations close to pi / 2 to ensure gradients are not too small
+        x = np.random.uniform(0.8, 2, size=qml.BasicEntanglerLayers.shape(n_layers=1, n_wires=3))
+
+        # for autograd in particular, take only the real part since it doesn't
+        # support complex differentiation
+        actual = qml.jacobian(lambda x: qml.math.real(shadow_circuit(x)))(x)
+        expected = qml.jacobian(lambda x: qml.math.real(exact_circuit(x)))(x)
+
+        assert qml.math.allclose(actual, expected, atol=1e-1)
+
+    @pytest.mark.jax
+    def test_backward_jax(self):
+        """Test the gradient of the state for the JAX interface"""
+        import jax
+        from jax import numpy as jnp
+
+        shadow_circuit = basic_entangler_circuit(3, shots=20000, interface="jax")
+        shadow_circuit = qml.shadows.state(wires=[0, 1, 2], diffable=True)(shadow_circuit)
+        exact_circuit = basic_entangler_circuit_exact_state(3, "jax")
+
+        # make rotations close to pi / 2 to ensure gradients are not too small
+        x = jnp.array(
+            np.random.uniform(
```

Comment:
you could alternatively define a global x with numpy (one) and then call this `x_jax = jnp.array(x)`

Reply:
This way you can compute one `expected` globally (or even hard-code it) so you dont have to repeat the exact computation for every backend.

Reply:
I can move the definition of `x` to the class scope to avoid duplication.

However, the `expected` would have to be computed in every test, since they run in different GitHub runners depending on the installed packages.

Reply:
Do you mean the different interfaces? In the end you can transform everything to numpy when you compare no?

Reply:
Yes, the different interfaces. I mean that we can't just compute the `expected` value only once, because each of the individual gradient tests run on different machines.

In fact, we're already computing the `expected` value only once for each test runner, so changing this would give no speedup.

Reply:
I dont understand, you can run
```python
exact_circuit = basic_entangler_circuit_exact_state(3, "jax")
expected = qml.math.real(jax.jacrev(exact_circuit, holomorphic=True)(x))
```
outside the test/class _once_ (here with jax, or alternatively with autograd), and then you can compare `actual` with `qml.math.convert_like(expected, actual)` in the test? So each test does not run the exact computation with different interfaces (which has to yield the same result if you fix one random but fix x).

Reply:
No we can't do that, because a TensorFlow test runner will not have JAX installed. And we can't use autograd to compute the gradients because it doesn't handle complex outputs well.

Reply:
Hm I am still puzzled by this, if you run the exact one outside and in autograd it should work no? Because each test will have the respective interface + autograd imported?

Reply:
Autograd doesn't handle complex gradients, so we can't use the computed values in autograd for the other interfaces.

Reply:
okay I see, easiest to leave it as is 👍